modules/util/src/main/native/android/c/Util.c

```diff
@@ -83,45 +84,40 @@ JavaVM* getGraalVM() {
     return graalVM;
 }
 
-///////////////////////////////////////////
-// native (Substrate) to native (Attach) //
-///////////////////////////////////////////
+//////////////////////////////////
+// native (Substrate) to Dalvik //
+//////////////////////////////////
 
 void attach_setActivityResult(jint requestCode, jint resultCode, jobject intent)
 {
-    if ((*androidVM)->GetEnv(androidVM, (void **)&androidEnv, JNI_VERSION_1_6) != JNI_OK) {
-        ATTACH_LOG_WARNING("attach_setActivityResult called from not-attached thread\n");
-        (*androidVM)->AttachCurrentThread(androidVM, (void **)&androidEnv, NULL);
-    }  else {
-        ATTACH_LOG_WARNING("attach_setActivityResult called from attached thread %p\n", androidEnv);
-    }
     ATTACH_LOG_FINE("call Attach::nativeDispatchActivityResult method from native: %d %d", requestCode, resultCode);
-    (*androidEnv)->CallStaticVoidMethod(androidEnv, jUtilClass, jUtilOnActivityResultMethod, requestCode, resultCode, intent);
+    ATTACH_DALVIK();
+    (*dalvikEnv)->CallStaticVoidMethod(dalvikEnv, jUtilClass, jUtilOnActivityResultMethod, requestCode, resultCode, intent);
+    DETACH_DALVIK();
     ATTACH_LOG_FINE("call Attach::nativeDispatchActivityResult method from native done");
 }
 
 void attach_setLifecycleEvent(const char* event) {
-    if ((*androidVM)->GetEnv(androidVM, (void **)&androidEnv, JNI_VERSION_1_6) != JNI_OK) {
-        ATTACH_LOG_WARNING("attach_setLifecycleEvent called from not-attached thread\n");
-        (*androidVM)->AttachCurrentThread(androidVM, (void **)&androidEnv, NULL);
-    }  else {
-        ATTACH_LOG_WARNING("attach_setLifecycleEvent called from attached thread %p\n", androidEnv);
-    }
-    ATTACH_LOG_FINE("call Attach method from native Lifecycle: %s", event);
-    jstring jchars = (*androidEnv)->NewStringUTF(androidEnv, event);
-    (*androidEnv)->CallStaticVoidMethod(androidEnv, jUtilClass, jUtilOnLifecycleEventMethod, jchars);
-    (*androidEnv)->DeleteLocalRef(androidEnv, jchars);
+    ATTACH_LOG_FINE("Call Attach method from native Lifecycle: %s", event);
+    ATTACH_DALVIK();
+    jstring jchars = (*dalvikEnv)->NewStringUTF(dalvikEnv, event);
+    ATTACH_LOG_FINE("Call2 Attach method from native Lifecycle: %s, class at %p and method at %p and jc at %p\n", event, jUtilClass, jUtilOnLifecycleEventMethod, jchars);
+    (*dalvikEnv)->CallStaticVoidMethod(dalvikEnv, jUtilClass, jUtilOnLifecycleEventMethod, jchars);
+    ATTACH_LOG_FINE("Call3 Attach method from native Lifecycle: %s", event);
+    (*dalvikEnv)->DeleteLocalRef(dalvikEnv, jchars);
+    ATTACH_LOG_FINE("Call4 Attach method from native Lifecycle: %s", event);
+    DETACH_DALVIK();
     ATTACH_LOG_FINE("called Attach method from native Lifecycle done");
 }
 
-///////////////////////////////////////
-// From native to Dalvik (Substrate) //
-///////////////////////////////////////
+////////////////////////////
+// From Dalvik to Dalvik  //
```

Comment:
Confusing :) Maybe keep which comes from Substrate?

Reply:
That is not relevant. The relevant thing is to understand the VM of the caller and callee: Graal or Dalvik. It doesn't matter if that comes via Substrate or Attach.